consensus/scheme/rolldpos/chainmanager.go

```diff
@@ -0,0 +1,426 @@
+package rolldpos
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/iotexproject/go-pkgs/crypto"
+	"github.com/iotexproject/go-pkgs/hash"
+	"github.com/iotexproject/iotex-address/address"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/zap"
+
+	"github.com/iotexproject/iotex-core/v2/action/protocol"
+	"github.com/iotexproject/iotex-core/v2/blockchain"
+	"github.com/iotexproject/iotex-core/v2/blockchain/block"
+	"github.com/iotexproject/iotex-core/v2/blockchain/genesis"
+	"github.com/iotexproject/iotex-core/v2/db"
+	"github.com/iotexproject/iotex-core/v2/pkg/log"
+	"github.com/iotexproject/iotex-core/v2/pkg/prometheustimer"
+)
+
+var (
+	blockMtc = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "iotex_blockmint_metrics",
+			Help: "Block mint metrics.",
+		},
+		[]string{"type"},
+	)
+)
+
+type (
+	// ChainManager defines the blockchain interface
+	ChainManager interface {
+		ForkChain
+		// Start starts the chain manager
+		Start(ctx context.Context) error
+		// Fork creates a new chain manager with the given hash
+		Fork(hash hash.Hash256) (ForkChain, error)
+		// CommitBlock validates and appends a block to the chain
+		CommitBlock(blk *block.Block) error
+		// ValidateBlock validates a new block before adding it to the blockchain
+		ValidateBlock(blk *block.Block) error
+	}
+	// ForkChain defines the blockchain interface
+	ForkChain interface {
+		// BlockProposeTime return propose time by height
+		BlockProposeTime(uint64) (time.Time, error)
+		// BlockCommitTime return commit time by height
+		BlockCommitTime(uint64) (time.Time, error)
+		// TipHeight returns tip block's height
+		TipHeight() uint64
+		// TipHash returns tip block's hash
+		TipHash() hash.Hash256
+		// StateReader returns the state reader
+		StateReader() (protocol.StateReader, error)
+		// MintNewBlock creates a new block with given actions
+		// Note: the coinbase transfer will be added to the given transfers when minting a new block
+		MintNewBlock(time.Time, crypto.PrivateKey, hash.Hash256) (*block.Block, error)
+	}
+	// StateReaderFactory is the factory interface of state reader
+	StateReaderFactory interface {
+		StateReaderAt(uint64, hash.Hash256) (protocol.StateReader, error)
+	}
+
+	// BlockBuilderFactory is the factory interface of block builder
+	BlockBuilderFactory interface {
+		Mint(ctx context.Context, pk crypto.PrivateKey) (*block.Block, error)
+		ReceiveBlock(*block.Block) error
+	}
+
+	chainManager struct {
+		ForkChain
```

Comment:
think use `fc ForkChain` is better
since `chainManager` implements all 6 methods of the `ForkChain` interface on its own